autokeras/blocks/basic.py

```diff
@@ -260,6 +261,222 @@ def _get_padding(kernel_size, output_node):
+        if num_heads is None:
+            num_heads = 8
+
+        if embed_dim % num_heads != 0:  # how to evaluate this condition
+            raise ValueError(
+                f"embedding dimension = {embed_dim} should be "
+                f"divisible by number of heads = {num_heads}"
+            )
+        projection_dim = embed_dim // num_heads
+        query_dense = layers.Dense(embed_dim)
+        key_dense = layers.Dense(embed_dim)
+        value_dense = layers.Dense(embed_dim)
+        combine_heads = layers.Dense(embed_dim)
+        batch_size = tf.shape(input_node)[0]
+        query = query_dense(input_node)  # (batch_size, seq_len, embed_dim)
+        key = key_dense(input_node)  # (batch_size, seq_len, embed_dim)
+        value = value_dense(input_node)  # (batch_size, seq_len, embed_dim)
+        query, key, value = [self.separate_heads(
+            var, batch_size, num_heads,  projection_dim
+        ) for var in [query, key, value]]
+        attention, weights = self.attention(query, key, value)
+        attention = tf.transpose(
+            attention, perm=[0, 2, 1, 3]
+        )  # (batch_size, seq_len, num_heads, projection_dim)
+        concat_attention = tf.reshape(
+            attention, (batch_size, tf.shape(attention)[1], self.embed_dim)
+        )  # (batch_size, seq_len, embed_dim)
+        output = combine_heads(
+            concat_attention
+        )  # (batch_size, seq_len, embed_dim)
+        return output
+
+    @staticmethod
+    def attention(query, key, value):
+        score = tf.matmul(query, key, transpose_b=True)
+        dim_key = tf.cast(tf.shape(key)[-1], tf.float32)
+        scaled_score = score / tf.math.sqrt(dim_key)
+        weights = tf.nn.softmax(scaled_score, axis=-1)
+        output = tf.matmul(weights, value)
+        return output, weights
+
+    @staticmethod
+    def separate_heads(x, batch_size, num_heads, projection_dim):
+        x = tf.reshape(x, (batch_size, -1, num_heads, projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+
+
+class Transformer(block_module.Block):
+    """Block for Transformer.
+    The input should be tokenized sequences with the same length, where each element
+    of a sequence should be the index of the word.
+    # Arguments
+        max_features: Int. Size of the vocabulary. Must be set if not using
+            TextToIntSequence before this block. Defaults to 20001.
+        pretraining: String. 'random' (use random weights instead any pretrained
+            model), 'glove', 'fasttext' or 'word2vec'. Use pretrained word embedding.
+            If left unspecified, it will be tuned automatically.
+        embedding_dim: Int. Output dimension of the Attention block.
+            If left unspecified, it will be tuned automatically.
+        num_heads: Int. The number of attention heads. If left unspecified,
+            it will be tuned automatically.
+        ff_dim: Int. The output dimension of the FFN. If left
```

Comment:
Always use fully-spelled out argument names instead of abbreviations, unless the abbreviation is already known to 100% of people who will read it.